triage/summarize.py

```diff
@@ -0,0 +1,369 @@
+        for j in range(len(t)):
+            v1[j + 1] = min(v1[j] + 1, v0[j + 1] + 1, v0[j] + int(s[i] != t[j]))
+
+        if limit and min(v1) > limit:
+            return min(v1)
+
+    return v1[len(t)]
+
+
+def make_ngram_counts(s, ngram_counts={}):
+    '''
+    Convert a string into a histogram of frequencies for different byte combinations.
+    This can be used as a heuristic to estimate edit distance between two strings in
+    constant time.
+    '''
+    size = 1 << 6
+    if s not in ngram_counts:
+        counts = [0] * size
+        for x in xrange(len(s)-3):
+            # print x, `s[x:x+4]`
+            counts[zlib.crc32(s[x:x+4].encode('utf8')) & (size - 1)] += 1
+        ngram_counts[s] = counts  # memoize
+    return ngram_counts[s]
+
+
+def make_ngram_counts_digest(s):
+    return hashlib.sha1(str(make_ngram_counts(s))).hexdigest()[:20]
+
+
+def ngram_editdist(a, b):
+    '''
+    Compute a heuristic edit distance using ngram counts.
+    '''
+    counts_a = make_ngram_counts(a)
+    counts_b = make_ngram_counts(b)
+    return sum(abs(x-y) for x, y in zip(counts_a, counts_b))/4
+
+
+def file_memoize(description, name):
+    def inner(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            if os.path.exists(name):
+                data = json.load(open(name))
+                print 'done (cached)', description
+                return data
+            else:
+                data = func(*args, **kwargs)
+                json.dump(data, open(name, 'w'))
+                print 'done', description
+                return data
+        wrapper.__wrapped__ = func
+        return wrapper
+    return inner
+
+
+@file_memoize('loading failed tests', 'failed.json')
+def load_failures(builds_file, tests_file):
+    builds = {}
+    for build in json.load(open(builds_file)):
+        if not build['started'] or not build['number']:
+            continue
+        for attr in ('started', 'tests_failed', 'number', 'tests_run'):
+            build[attr] = int(build[attr])
+        build['elapsed'] = int(float(build['elapsed']))
+        if 'pr-logs' in build['path']:
+            build['pr'] = build['path'].split('/')[-3]
+        builds[build['path']] = build
+
+    failed_tests = {}
+    for test in json.load(open(tests_file)):
+        failed_tests.setdefault(test['name'], []).append(test)
+    for tests in failed_tests.itervalues():
+        tests.sort(key=lambda t: t['build'])
+
+    return builds, failed_tests
+
+
+def cluster_test(tests):
+    '''
+    Compute failure clusters given a list of failures for one test.
+
+    Args:
+        tests: list of failed test dictionaries, with 'failure_text' keys
+    Returns:
+        {failure_text: [failure_in_cluster_1, failure_in_cluster_2, ...]}
+    '''
+    clusters = {}
+
+    for test in tests:
+        ftext = test['failure_text']
+        fnorm = normalize(ftext)
+        if fnorm in clusters:
+            clusters[fnorm].append(test)
+        else:
```

Comment:
I think if/else is clearer to read than a continue-- less non-local context required. The indentation doesn't isn't very deep here, regardless.